src/modules/xine/link_deinterlace.c

```diff
@@ -0,0 +1,331 @@
+		int progressive_source = mlt_properties_get_int( MLT_PRODUCER_PROPERTIES(original_producer), "progressive" );
+		pdata->scan_mode_detected = 1;
+		char *scan_mode = progressive_source ? "progressive" : "interlaced";
+		mlt_log_info( MLT_LINK_SERVICE(self), "Scan mode detected: %s\n", scan_mode );
+		if ( progressive_source )
+		{
+			mlt_log_info( MLT_LINK_SERVICE(self), "Do not deinterlace - progressive source\n" );
+			pdata->deinterlace_required = 0;
+			mlt_producer_prepare_next( MLT_LINK_PRODUCER( self ) );
+			return error;
+		}
+		else
+		{
+			// Proceed to setup the frame cache below
+			mlt_frame_close(*frame);
+		}
+	}
+
+	if ( !pdata->deinterlace_required )
+	{
+		// As soon as get_image() detects that deinterlace is not required we stop passing
+		// future frames and stop calling get_image().
+		// This can not be undone. So a source can not toggle between interlaced/progressive.
+		mlt_producer_seek( self->next, frame_pos );
+		error = mlt_service_get_frame( MLT_PRODUCER_SERVICE( self->next ), frame, index );
+		mlt_producer_prepare_next( MLT_LINK_PRODUCER( self ) );
+		return error;
+	}
+
+	mlt_producer_seek( self->next, frame_pos );
+	error = mlt_service_get_frame( MLT_PRODUCER_SERVICE( self->next ), frame, index );
+	if ( error )
+	{
+		mlt_log_error( MLT_LINK_SERVICE(self), "Unable to get frame: %d\n", frame_pos );
+		return error;
+	}
+
+	if ( pdata->prev_next_required )
+	{
+		mlt_properties unique_properties = mlt_frame_unique_properties( *frame, MLT_LINK_SERVICE(self) );
+		mlt_producer_seek( self->next, frame_pos - 1 );
+		error = mlt_service_get_frame( MLT_PRODUCER_SERVICE( self->next ), &prev, index );
+		if ( error )
+		{
+			mlt_log_error( MLT_LINK_SERVICE(self), "Unable to get prev: %d\n", frame_pos );
+		}
+		mlt_properties_set_data( unique_properties, "prev", prev, 0, (mlt_destructor)mlt_frame_close, NULL );
+
+		mlt_producer_seek( self->next, frame_pos + 1 );
+		error = mlt_service_get_frame( MLT_PRODUCER_SERVICE( self->next ), &next, index );
+		if ( error )
+		{
+			mlt_log_error( MLT_LINK_SERVICE(self), "Unable to get next: %d\n", frame_pos );
+		}
+		mlt_properties_set_data( unique_properties, "next", next, 0, (mlt_destructor)mlt_frame_close, NULL );
+	}
+
+	mlt_frame_push_service( *frame, self );
+	mlt_frame_push_get_image( *frame, link_get_image );
+	mlt_producer_prepare_next( MLT_LINK_PRODUCER( self ) );
+
+	// This link will request the native image. Add normalizers to be applied after this link.
+	if ( !pdata->rescale_filter )
+	{
+		pdata->rescale_filter = mlt_factory_filter( mlt_service_profile( MLT_LINK_SERVICE( self ) ), "swscale", NULL );
+	}
+	if ( pdata->rescale_filter )
+	{
+		mlt_filter_process( pdata->rescale_filter, *frame );
+	}
+	if ( !pdata->resize_filter )
+	{
+		pdata->resize_filter = mlt_factory_filter( mlt_service_profile( MLT_LINK_SERVICE( self ) ), "resize", NULL );
+	}
+	if ( pdata->resize_filter )
+	{
+		mlt_filter_process( pdata->resize_filter, *frame );
+	}
```

Comment:
It is not good that a link needs to codify the normalizing filters that the loader producer is responsible for. What happens with crop and/or movit?

Reply:
I agree. It is an unfortunate work-around.

Movit was not a factor when I started this project - I did not know you were going to resurrect it :)

Crop Source still works since there is nothing defeating the crop and pad normalizers.

Maybe I should prioritize making corresponding normalizing links. They would just be very simple wrappers around the existing filters.